### PR TITLE
[Fix]jsonrpc getting consistent song and album artist data - names, IDs and MusicbrainzIDs

### PR DESCRIPTION
**Pull request by DaveTBlake**
This fixes the issue that JSON may in some circumstances return inconsistent artist name and ID arrays for song and albums. It also correctly provides Musicbrainz IDs, that would never have been working properly before.

Since adding processing for ARTISTS tag #7486 it is possible that the splitting the artists string may not give the same artists as held in the song_artist or album_artist tables, so could lead to discrepencies. In particular `audioLibrary::GetAlbums` and `GetSongs` the array of artists for an album may not match the array of artistIds.  This was discussed under #8268 

It ensures that when JSON gets albums or songs the underlying query uses song_artist or album_artist tables to fully populate artist credits. Hence artist names and id arrays will be consistent. As a bonus now gets MBID (as a vector to match names and ids)

It bumps JSON version for now artist data returned is consistent.

This is a fix, not an improvement, but does make changes to underlying queries so could be considered major. I have deliberately avoided impacting the main music navigation SQL, but longer term we should be getting artist data from the song_artist table in more places.

Even if it does not make Jarvis alpha since it is JSON related @Montellese I am hoping you can have a look at this before you are away for a month, and at least approve the approach. It needs some more testing with live data, so I am hoping a team member will build it so this can happen.

@Tolriq see I did as I promised. Hope that you can test it with some Yatse data.
@Razzeee and @evilhamster have a look please since it is music.

### CODE REVIEW

File: xbmc/music/Album.cpp [hunk 1]
```diff
@@ -181,8 +181,8 @@ const std::vector<std::string> CAlbum::GetAlbumArtist() const
   //When artist credits have not been populated attempt to build an artist vector from the descrpition string
   //This is a tempory fix, in the longer term other areas should query the album_artist table and populate
   //artist credits. Note that splitting the string may not give the same artists as held in the album_artist table
-  if (albumartists.empty() && !strArtistDesc.empty())
-    albumartists = StringUtils::Split(strArtistDesc, g_advancedSettings.m_musicItemSeparator);
+//  if (albumartists.empty() && !strArtistDesc.empty())
+//    albumartists = StringUtils::Split(strArtistDesc, g_advancedSettings.m_musicItemSeparator);
```
Comment (Razzeee) [inline xbmc/music/Album.cpp]: Is this ment to be commented out? Either you forget to remove it or work on it some more, I guess?

File: xbmc/music/MusicDatabase.cpp [hunk 2]
```diff
@@ -1987,12 +2022,14 @@ bool CMusicDatabase::GetTop100Albums(VECALBUMS& albums)
...
 
-    // NOTE: The song.idAlbum is needed for the group by, as for some reason group by albumview.idAlbum doesn't work
-    //       consistently - possibly an SQLite bug, as it works fine in SQLiteSpy (v3.3.17)
-    std::string strSQL = "select albumview.* from albumview "
-                    "where albumview.iTimesPlayed>0 and albumview.strAlbum != '' "
-                    "order by albumview.iTimesPlayed desc "
-                    "limit 100 ";
+    // Get data from album and album_artist tables to fully populate albums
+    std::string strSQL = "SELECT albumview.*, albumartistview.* FROM albumview "
+      "LEFT JOIN albumartistview ON albumview.idAlbum = albumartistview.idAlbum "
```
Comment (Razzeee) [inline xbmc/music/MusicDatabase.cpp]: Not sure why this is a left join. I don't think albumartistview can have data that doesn't match (haven't tested, but we should check this). If it has, do we have logic in place to handle with the NULL values that will be thrown our way?

File: xbmc/music/MusicDatabase.h [hunk 3]
```diff
@@ -349,12 +349,14 @@ class CMusicDatabase : public CDatabase
   bool GetCommonNav(const std::string &strBaseDir, const std::string &table, const std::string &labelField, CFileItemList &items, const Filter &filter /* = Filter() */, bool countOnly /* = false */);
   bool GetAlbumTypesNav(const std::string &strBaseDir, CFileItemList &items, const Filter &filter = Filter(), bool countOnly = false);
   bool GetMusicLabelsNav(const std::string &strBaseDir, CFileItemList &items, const Filter &filter = Filter(), bool countOnly = false);
-  bool GetAlbumsNav(const std::string& strBaseDir, CFileItemList& items, int idGenre = -1, int idArtist = -1, const Filter &filter = Filter(), const SortDescription &sortDescription = SortDescription(), bool countOnly = false);
+  bool GetAlbumsNav(const std::string& strBaseDir, CFileItemList& items, int idGenre = -1, int idArtist = -1, const Filter &filter = Filter(), const SortDescription &sortDescription = SortDescription(), bool countOnly = false);
```
Comment (Razzeee) [inline xbmc/music/MusicDatabase.h]: Unneeded whitespace change.

File: xbmc/music/Song.cpp [hunk 4]
```diff
@@ -165,8 +165,8 @@ const std::vector<std::string> CSong::GetArtist() const
   //When artist credits have not been populated attempt to build an artist vector from the descrpition string
   //This is a tempory fix, in the longer term other areas should query the song_artist table and populate
   //artist credits. Note that splitting the string may not give the same artists as held in the song_artist table
-  if (songartists.empty() && !strArtistDesc.empty())
-    songartists = StringUtils::Split(strArtistDesc, g_advancedSettings.m_musicItemSeparator);
+  //if (songartists.empty() && !strArtistDesc.empty())
+  //  songartists = StringUtils::Split(strArtistDesc, g_advancedSettings.m_musicItemSeparator);
```
Comment (Razzeee) [inline xbmc/music/Song.cpp]: More commented out code. I'm not really a fan. I know this from work, usually you forget about it and half a year later someone looks at it and has no clue why it's commented out. :/

File: xbmc/music/Album.cpp (hunk 1 above)
Comment (DaveTBlake) [inline xbmc/music/Album.cpp]: I introduced those lines as a temp fix a week ago, as the comment above them says. Here I am trying with them out. I think I have fixed the issue, but I want to get some real data testing going to be absolutely sure. Don't want to make the same mistake twice!

File: xbmc/music/MusicDatabase.cpp (hunk 2 above)
Comment (DaveTBlake) [inline xbmc/music/MusicDatabase.cpp]: Left join because we can have albums without artists, and we still want to see them. It was discussed in another PR, and @Tolriq confirmed that we do have albums and songs without artists sometimes.  An abberation but could happen with tags as data source.  I don't see absence of artist causing an issue here.

File: xbmc/music/Song.cpp (hunk 4 above)
Comment (DaveTBlake) [inline xbmc/music/Song.cpp]: Not something I would normally do, and I would expect to remove before merge. There is just a doubt in my mind that some other part of Kodi is using the description string to  make artists. By commenting out it will show up in testing. I really want this built so others can test it, and @Montellese to have some chance to look at it.  Not suggesting we merge before that has happened.

File: xbmc/music/Album.cpp (hunk 1 above)
Comment (Razzeee) [inline xbmc/music/Album.cpp]: Sure. Don't get me wrong, but let's just get rid of them and probably the comments, if it doesn't apply anymore. After all this code is replacing/improving it.

File: xbmc/music/MusicDatabase.cpp (hunk 2 above)
Comment (Razzeee) [inline xbmc/music/MusicDatabase.cpp]: Weird, case but then we can't do anything about that. Thanks, for checking.

File: xbmc/music/Song.cpp (hunk 4 above)
Comment (Razzeee) [inline xbmc/music/Song.cpp]: I'll do tests build asap. (Need read up on how)

**Comment by kib** [comment]
@DaveTBlake @Razzeee The builds are finished, linked to them here http://forum.kodi.tv/showthread.php?tid=245312&pid=2145207#pid2145207

**Comment by DaveTBlake** [comment]
Thanks for the builds @kib 
I have had some testing feedback from @Tolriq. Don't know if @ronie would like to look at this having caught previous issues with GetAlbums. 
@Montellese any chance of a comment before you go? Would like to merge this while you are away, so a thumbs up from you (even with reservations if you have not had time to test) would be good.

**Comment by ronie** [comment]
will give it a spin in the next few days.

**Comment by Montellese** [comment]
You don't alter the API definition so nothing that can go wrong there. Otherwise I'm not really into how the music library and that musicbrainz stuff works at all so not much I can review.

**Comment by ronie** [comment]
tested this PR and didn't spot any breakage.

i did however (off topic to this PR) notice an issue with the 'artist mood' and 'album mood' field as returned by json. previously it was returned as a list: [mood1, mood2, mood3]
but now it's returned as a string: u'mood1 / mood2 / mood3'

**Comment by Tolriq** [comment]
According to JSON definition this have not changed and should still be an array.

My code is change proof for those since so many changes in previous versions. Did not notice this one.

No time this week end but maybe 
```    "instrument": { "$ref": "Array.String" },
      "style": { "$ref": "Array.String" },
      "mood": { "$ref": "Array.String" },
      "yearsactive": { "$ref": "Array.String" },
```
Should be checked too

**Comment by DaveTBlake** [comment]
@ronie, @Tolriq what has changed in Mood isn't anything to do with this PR, but I will have a dig and see if I can spot what and when. CMusicDatabase::SetPropertiesFromArtist has both Mood and Mood_Array properties.

**Comment by DaveTBlake** [comment]
I am certain that the changes you are seeing with mood are down to #8012 . It would seem @Montellese inadvertently impacted the interface format.

Just considering albums for a moment, the mood data is in the album table as a single string with "/" separators. It is read into CAlbum split into a vector of strings. Then CMusicDatabase::SetPropertiesFromAlbum stores it in "album_mood"  and "album_mood_array" properties (outside the item.MusicInfoTag structure). But CFileItem::SetFromAlbum calls both SetPropertiesFromAlbum and GetMusicInfoTag()->SetAlbum(album). The latter is changed by #8012 to set the mood string CMusicInfoTag.m_strMood, so now we have mood in 2 places in CFileItem.  Subsequently CFileItemHandler::HandleFileItem calls FillDetails calls GetField that first tries to get mood from serialised value of CMusicInfoTag ("mood1 / mood2 /...") and then if that is empty looks in property "album_mood_array" ("Mood1", "Mood2" etc.) and finally if that is empty too tries "album_mood" ("mood1 / mood2 /..."). 

Bottom line is I understand why you are seeing what you are. Question is what, if anything, do we do about it?

**Comment by Tolriq** [comment]
Well there's need to fix things in one way or another :(

Would be @Montellese choice as I have no big picture vision. But currently if returned value is a String and JSON definition says it's an Array there's an API error and either mood should always return a String and we update the JSON definition or we keep current definition (best IMO as earlier most concatened String returns where switched to Array so I suppose there's no will to return to concatened Strings.) and ensure that the field is always an array by splitting the String if not an Array, or ensuring correct array data is always used / set.

**Comment by DaveTBlake** [comment]
Mood fix is another PR
This PR @Razzeee has checked over the code (@evilahamster seems to be away). @Montellese had no comment to make before going on holiday since it is music and doesn't change API. @Ronie and @Tolriq  have tested with real data. I think we are good to go :)

Jenkins build and merge please.

**Comment by DaveTBlake** [comment]
Jenkins build and merge

**Comment by jjd-uk** [comment]
jenkins build this please

**Comment by hudokkow** [comment]
Welcome and congrats on your first merge. :wink:

**Comment by Tolriq** [comment]
@DaveTBlake since I see the PR for bump for Beta 1 is already at the corner let's try to not forget for release :(
